feat(net): test syncing from geth

Rjected opened the pull request:
## reth integration tests

This adds tests which spawn an instance of `geth` and configure it to produce blocks with cilque. Geth is sent transactions, which gets used to produce blocks. This is then used to test the following situations:
 * Reth full sync test, adding geth as reth's only outbound peer
 * Reth network keepalive test, initializing just reth's networking stack, and making sure we can stay connected to a peered geth instance for 30 seconds.

Depends on:
 - [x] https://github.com/gakonst/ethers-rs/pull/1975
 - [x] https://github.com/gakonst/ethers-rs/pull/2008
 - [x] https://github.com/gakonst/ethers-rs/pull/2009
 - [x] https://github.com/gakonst/ethers-rs/pull/2010
 - [x] #939
 - [x] https://github.com/gakonst/ethers-rs/pull/2068
 - [x] #965
 - [x] https://github.com/gakonst/ethers-rs/pull/2091
 - [x] https://github.com/gakonst/ethers-rs/pull/2093

TODO:
 - [x] Write sync test that sets up the reth pipeline and peers geth
   - [x] Get geth to produce blocks
   - [x] Set up a reth `Genesis`
   - [x] Rest of pipeline
   - [x] Put transactions in blocks (currently they are empty but we do have a funded account)
 - [x] Consider moving to different folder since we will be using the entire pipeline
 - [x] Figure out how to produce blocks
    - Configures geth in with clique, sending transactions that should get automatically mined.
 - [x] Remove patch
 - [x] Integrate pipeline builder from #1017

gakonst commented:
@Rjected gm - planned next steps here?

Rjected commented:
gm @gakonst 

still debugging an invalid sender error, looking through geth to see if I activated EIP155 correctly

Then:
 - [ ] Initialize other components
 - [ ] Run the test

gakonst commented:
Cool - pls ping me if you're still blocked on this by eod (if you're online today?) so that we debug together as it's important we get confidence that it's the network and not us causing the peering issues

mattsse reviewed (commented):
this looks great!
incredibly useful.

should we go over this after the call Monday to assess what's missing/blockers etc?

Rjected commented:
> should we go over this after the call Monday to assess what's missing/blockers etc?

yeah that would be very useful, lets do it @mattsse

Rjected commented:
big refactor:
 * network test utils are now in a new crate `reth-net-test-utils`. I am thinking of bringing this out into a module `reth_network::test_utils` (like `reth_provider::test_utils`), as long as it doesn't add unnecessary dependencies to the `network` crate. additional dev-deps are fine
 * the geth sync test now lives in `reth-tests` (in `crates/tests`), which will contain other structs and utils for writing integration tests with geth. we can also import other parts of reth from here

Rjected commented:
almost done, just noting down things left TODO:
 - [ ] Create conversions from `ethers_core::utils::ChainConfig` => `reth_consensus::Config`, then use that to initialize `BeaconConsensus` and pass it into the `RethBuilder`
 - [ ] Create conversion from `ethers_core::utils::Genesis` => `reth_cli_utils::Genesis`, also pass it into `RethBuilder`
 - [ ] Fill blocks with transactions and change period to `0` so blocks seal instantly
 - [ ] Assert that we get blocks, headers, etc

gakonst commented on crates/net/network/Cargo.toml:
```diff
@@ -69,9 +69,12 @@ reth-interfaces = { path = "../../interfaces", features = ["test-utils"] }
 reth-provider = { path = "../../storage/provider", features = ["test-utils"] }
 reth-tracing = { path = "../../tracing" }
 reth-transaction-pool = { path = "../../transaction-pool", features = ["test-utils"] }
+reth-net-test-utils = { path = "../test-utils" }
```
Do we need this to be a separate crate?

Rjected replied:
we don't need this to be a separate crate, we can make it a submodule of `network`

gakonst commented on crates/net/test-utils/src/testnet.rs:
```diff
@@ -52,22 +52,28 @@ where
         &mut self.peers
     }
 
+    /// Return a slice of all peers.
```
I would prefer all these to remain under net/ and have them exported via a test-utils feature?

Let's do these as smaller refactors over the entire PR

Rjected replied:
that makes sense, we can create a `network::test_utils` module under a `test-utils` feature

gakonst commented on crates/tests/Cargo.toml:
```diff
@@ -0,0 +1,55 @@
+[package]
+name = "reth-tests"
```
Most of these feel like the should be in a reth library which is then imported by the CLI?

gakonst commented on crates/tests/src/clique.rs:
```diff
@@ -0,0 +1,374 @@
+use std::collections::HashMap;
+
+/// Converts an ethers [`Genesis`] into a reth
+/// [`ChainSpecification`](reth_cli_utils::chainspec::ChainSpecification).
```
This should probably be a From implementation and the ChainSpecification should be part of the reth crate I mentioned

gakonst commented on crates/tests/src/clique.rs:
```diff
@@ -0,0 +1,374 @@
+}
+
+/// Extracts the genesis block header from an ethers [`Genesis`](ethers_core::utils::Genesis).
+pub(crate) fn genesis_header(genesis: &Genesis) -> Header {
```
Similar here re Genesis type. I personally don't mind depending on ethers for some things like that.

Rjected replied:
sounds good, will integrate all conversions as `From` impls

gakonst commented on crates/tests/src/sync.rs:
```diff
@@ -0,0 +1,227 @@
+    // first get the balance and make sure its not zero
+    let balance = provider.get_balance(our_address, None).await.unwrap();
+    assert_ne!(balance, 0u64.into());
+    println!("address: {our_address:?}");
+    println!("balance at genesis: {balance:?}");
```
Remove these or make tracing logs

gakonst commented on crates/tests/src/sync.rs:
```diff
@@ -0,0 +1,227 @@
+        assert!(block > U64::zero());
+
+        // get the tip so we can send it to reth
+        let tip_block = provider.get_block(BlockNumber::Latest).await.unwrap().unwrap();
+        let tip_hash = tip_block.hash.unwrap().0.into();
```
All of this setup code should be  on a helper, probably on the test clique geth struct

gakonst commented on crates/tests/src/sync.rs:
```diff
@@ -0,0 +1,227 @@
+        // initialize consensus
+        let consensus = Arc::new(BeaconConsensus::new(chainspec.consensus));
+
+        // build reth and start the pipeline
+        let reth: RethTestInstance<Env<WriteMap>> = RethBuilder::new()
+            .db(db)
+            .consensus(consensus)
+            .genesis(chainspec.genesis)
+            .network(handle.clone())
+            .tip(tip_hash)
+            .build();
```
I like that we're testing syncing. Would have also liked some way to test p2p in isolation without running the entire pipeline inside the net crate

Rjected replied:
We already have this as [`test_incoming_connect_with_single_geth`](https://github.com/paradigmxyz/reth/blob/27ed772d31b777882c6f1d1173b20609d3450624/crates/net/network/tests/it/connect.rs#L376) and a few other tests, although in those cases geth doesn't have any blocks. We could add to those with a test using the clique-configured geth, but am curious if you had anything in mind

gakonst reviewed (changes requested):
I like where this landed. A big nitpick but I think we need some refactors and splitting in smaller PRs because it's a lot. I also feel we don't need dedicated crates for these and should be able to roll most in the existing ones?

Rjected commented:
squashed the commits and rebased on main - this now only contains code related to `crates/tests`

The only thing blocking this PR is a `CliqueConsensus` implementation, because `BeaconConsensus` enforces that `extraData` cannot be longer than 32 bytes, and clique always produces `extraData` longer than 32 bytes, which causes the geth peer to be kicked after failing header validation.
As a result, this integration test would not be covering `BeaconConsenus` validation.

mattsse commented on crates/staged-sync/tests/sync.rs:
```diff
@@ -0,0 +1,310 @@
+        drop(clique);
+
+        // cleanup (delete the data_dir at dir_path)
+        fs::remove_dir_all(dir_path).await.unwrap();
+    })
+    .await
+    .unwrap();
```
I believe this should be unnecessary because tempdir already does this on drop

Rjected replied:
ah I had to do this because I was using `into_path`, which requires manual deletion, but switching to `path` works and does not require doing this

mattsse commented on crates/staged-sync/tests/sync.rs:
```diff
@@ -0,0 +1,310 @@
+        );
+
+        // cleanup (delete the data_dir at dir_path)
+        fs::remove_dir_all(dir_path).await.unwrap();
+    })
+    .await
+    .unwrap();
```
same

mattsse commented on crates/staged-sync/src/test_utils/clique.rs:
```diff
@@ -0,0 +1,120 @@
+        // print logs in a new task
+        let mut err_reader = BufReader::new(stderr);
+
+        tokio::spawn(async move {
+            loop {
+                if let (Ok(line), _line_str) = {
+                    let mut buf = String::new();
+                    (err_reader.read_line(&mut buf), buf.clone())
+                } {
+                    if line == 0 {
+                        break
+                    }
+                }
+            }
```
this is just for debugging, right?

Rjected replied:
`print_logs` is for debugging, but `prevent_blocking` is necessary to prevent the stderr pipe buffer from filling up and causing the test to hang. Using `Drop` doesn't seem to work here because it causes geth to crash (probably due to `EPIPE`, but I'm not entirely sure). I'm not sure if there's a better way to prevent the stderr pipe buffer from filling though

Rjected replied:
we can't do it by default because `Geth` captures the stderr to make sure rpc, p2p, etc are up when `spawn` is called. This means we automatically pipe the `Command`'s output with `Stdio::piped()`. Maybe we could spawn `cat` or something, with geth's stderr as `cat`'s `stdin`, and pipe cat to /dev/null

mattsse commented on crates/staged-sync/src/lib.rs:
```diff
@@ -2,3 +2,7 @@ pub mod config;
 pub use config::Config;
 
 pub mod utils;
+
+#[cfg(any(test, feature = "test-utils"))]
+/// Common helpers for integration testing.
+pub mod test_utils;
```
I wonder if staged-sync is the right crate here, I guess so we can use the pipeline and test certain parts of it against a GethInstance?

perhaps it would be appropriate to move to its own crate?

wdyt @onbjerg ?

mattsse reviewed (changes requested):
After tons of ether PRs and debugging, this will be very useful.
great work!

only smol nit.
and we need to discuss where to put this exactly.

gakonst commented on crates/staged-sync/tests/sync.rs:
```diff
@@ -0,0 +1,319 @@
+        handle.add_peer(geth_peer_id, geth_socket);
+
+        // wait for the session to be established
+        let _peer_id = events.peer_added_and_established().await.unwrap();
```
shouldn't we add a `assert_eq!(geth_peer_id, peer_id)` here? what is the purpose of the keepalive stuff below?

gakonst reviewed (approved):
finally :) this exposed a lot of things to improve, so im glad we did this.
